### PR TITLE
feat(cli-summary-report): Create summary reports (html summary and error log)

### PR DESCRIPTION
**Pull request by AhmedAbdoOrtiga**
#### Description of changes

- Integrate summary report creation into crawler.
- Create error log.
- Update CLI notification

#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Added relevant unit test for your changes. (`yarn test`)
- [x] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [x] Ran precheckin (`yarn precheckin`)

### CODE REVIEW

File: packages/cli/src/report/report-disk-writer.ts [hunk 1]
```diff
@@ -35,4 +36,28 @@ export class ReportDiskWriter {
 
         return reportFileName;
     }
+
+    public writeErrorLogToDirectory(directory: string, fileName: string, errors: PageError[]): void {
+        if (isEmpty(directory)) {
+            // tslint:disable-next-line: no-parameter-reassignment
+            directory = '.';
```
Comment (lamaks) [inline packages/cli/src/report/report-disk-writer.ts:43]: use __dirname node var instead

File: packages/cli/src/report/report-disk-writer.ts [hunk 2]
```diff
@@ -35,4 +36,28 @@ export class ReportDiskWriter {
...
+        }
+
+        if (!this.fileSystemObj.existsSync(directory)) {
+            console.log('output directory does not exists.');
+            console.log(`creating output directory - ${directory}`);
```
Comment (lamaks) [inline packages/cli/src/report/report-disk-writer.ts:48]: no sure we need this logs

File: packages/cli/src/report/report-generator.spec.ts [hunk 3]
```diff
@@ -59,4 +65,45 @@ describe('ReportGenerator', () => {
...
+            .verifiable(Times.once());
+
+        reporterMock
+            .setup((rm) => rm.fromSummaryResults(parameters))
+            .returns(() => htmlReport)
+            .verifiable(Times.once());
+
+        const report = await reportGenerator.generateSummaryReport(crawlDetails, results);
+
+        reporterMock.verifyAll();
```
Comment (lamaks) [inline packages/cli/src/report/report-generator.spec.ts:106]: move to afterEach() and add verifyAll() for axeInfoMock, userAgentInfoMock

File: packages/cli/src/runner/crawler-command-runner.ts [hunk 4]
```diff
@@ -1,17 +1,31 @@
...
+        @inject(CrawlerEntryPoint) private readonly crawlerEntryPoint: CrawlerEntryPoint,
+        @inject(ReportGenerator) private readonly reportGenerator: ReportGenerator,
+        @inject(ReportDiskWriter) private readonly reportDiskWriter: ReportDiskWriter,
+        @inject(ReportNameGenerator) private readonly reportNameGenerator: ReportNameGenerator,
+    ) {}
 
     public async runCommand(scanArguments: ScanArguments): Promise<void> {
-        await this.crawlerEntryPoint.crawl({
+        const startDate = new Date();
+        const startDatNumber = Date.now();
```
Comment (lamaks) [inline packages/cli/src/runner/crawler-command-runner.ts:24]: typo

File: packages/cli/src/runner/crawler-command-runner.ts [hunk 5]
```diff
@@ -25,5 +39,34 @@ export class CrawlerCommandRunner implements CommandRunner {
...
+        };
+
+        const reportContent = await this.reportGenerator.generateSummaryReport(crawlDetails, scanResult.summaryScanResults);
+
+        const reportLocation = this.reportDiskWriter.writeToDirectory(scanArguments.output, 'index', 'html', reportContent);
+        console.log(`Summary report was saved as ${scanArguments.output}\\${reportLocation}`);
+
+        const errorLogLocation = `${this.reportNameGenerator.generateName('ai-cli-errors', endDate)}.log`;
+        this.reportDiskWriter.writeErrorLogToDirectory(scanArguments.output, errorLogLocation, scanResult.errors);
+        console.log(`Error log was saved as ${scanArguments.output}\\${reportLocation}`);
```
Comment (lamaks) [inline packages/cli/src/runner/crawler-command-runner.ts:70]: will not work on linux. Use filenamify to convert path

File: packages/crawler/src/level-storage/data-base.ts [hunk 6]
```diff
@@ -91,7 +92,15 @@ export class DataBase {
...
-        return { errors: errors, summaryScanResults: { failed: failed, passed: passed, unscannable: browserErrors } };
+        let promise = Promise.resolve(undefined);
+        await promise;
+
+        promise = new Promise((fulfill) => stream.on('end', fulfill));
+        await promise;
+
+        promise = Promise.resolve({ errors: errors, summaryScanResults: { failed: failed, passed: passed, unscannable: browserErrors } });
+
+        return promise;
```
Comment (lamaks) [inline packages/crawler/src/level-storage/data-base.ts:103]: Should be similar to:

```
await new Promise((fulfill) => stream.on('end', fulfill));
return { errors: errors, summaryScanResults: { failed: failed, passed: passed, unscannable: browserErrors } };
```

File: packages/cli/src/tool-data/user-agent-info.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        const browser = await this.puppeteer.launch({ args: ['--disable-dev-shm-usage'] });
+
+        // tslint:disable-next-line:no-unnecessary-local-variable
+        const userAgent = await browser.userAgent();
+
+        await browser.close();
+
+        return userAgent;
+    }
+}
```
Comment (lamaks) [inline packages/cli/src/tool-data/user-agent-info.ts:21]: This is not efficient to launch browser just to get user agent string. We need to get it from where we run scan, but not from UserAgentInfo. No need to have this class.

File: packages/cli/src/runner/crawler-command-runner.ts (hunk 5 above)
Comment (AhmedAbdoOrtiga) [reply packages/cli/src/runner/crawler-command-runner.ts:70]: I'll use 'path' package

File: packages/cli/src/runner/crawler-command-runner.ts [hunk 8]
```diff
@@ -65,8 +65,8 @@ export class CrawlerCommandRunner implements CommandRunner {
         const reportLocation = this.reportDiskWriter.writeToDirectory(scanArguments.output, 'index', 'html', reportContent);
         console.log(`Summary report was saved as ${scanArguments.output}\\${reportLocation}`);
```
Comment (lamaks) [inline packages/cli/src/runner/crawler-command-runner.ts:66]: add `path` here too to build agnostic path

**Review by lamaks — APPROVED** [review]

**Pull request merged**